clang/lib/Sema/SemaDeclCXX.cpp

```diff
@@ -19109,7 +19119,17 @@ bool Sema::DefineUsedVTables() {
         }
       }
 
-      if (IsExplicitInstantiationDeclaration)
+      if (IsExplicitInstantiationDeclaration &&
+          llvm::none_of(Class->decls(), [](Decl *decl) {
+            // If the class has a virtual member function declared with
+            // `__attribute__((exclude_from_explicit_instantiation))`, the
+            // explicit instantiation declaration shouldn't suppress emitting
+            // the vtable to ensure that the excluded member function is
+            // accessible through the vtable.
+            auto *Method = dyn_cast<CXXMethodDecl>(decl);
+            return Method && Method->isVirtual() &&
+                   Method->hasAttr<ExcludeFromExplicitInstantiationAttr>();
+          }))
```

Comment:
As a style issue, I think it would be easier to read if we put this in a variable, like:

```
bool HasExcludeFromExplicitInstantiation = llvm::any_of(... hasAttr<ExcludeFromExlicitInstantiationAttr>() ...
```

Also, I'm not sure that it matters whether the method with the attribute is virtual or not. I think it would be enough to check that *any* method has the attribute -- because then the motivation about "suppress the vtable; it will live with the explicit instantiation definition" may not hold.

Reply:
I wonder if we could break this out into a separate PR as well, since it seems somewhat separate from the other issue.

Reply:
> I don't understand why non-virtual functions matter here.

A constructor method may trigger emitting the vtable. The larger point is that the reasoning on lines 19095-190198 about the "vtable will live with the explicit instantiation" doesn't necessarily hold if some methods are excluded from explicit instantiation.

Reply:
> Why are non-virtual functions allowed to affect vtables?

When there's no key function, emitting a constructor will trigger emitting the vtable, because the constructor needs to set the vptr to point to it.

That's why the reasoning above about "vtable will live with the explicit instantiation" won't hold if the constructor is excluded.